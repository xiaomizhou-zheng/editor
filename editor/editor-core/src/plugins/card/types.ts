import { INPUT_METHOD, ACTION } from '../analytics/types/enums';
import {
  CardProvider,
  CardAppearance,
  CardAdf,
} from '@atlaskit/editor-common/provider-factory';
import { SmartLinkEvents } from '@atlaskit/smart-card';
import { CreateUIAnalyticsEvent } from '@atlaskit/analytics-next';
import { CardOptions } from '@atlaskit/editor-common/card';
import { LinkPickerOptions } from '../hyperlink/types';

export type CardInfo = {
  title?: string;
  url?: string;
  pos: number;
};

export type Request = {
  pos: number;
  url: string;
  appearance: CardAppearance;
  compareLinkText: boolean;
  source: CardReplacementInputMethod;
  analyticsAction?: ACTION;
  shouldReplaceLink?: boolean;
};

export type OutstandingRequests = { [key: string]: Promise<void | CardAdf> };

export type CardPluginState = {
  requests: Request[];
  provider: CardProvider | null;
  cards: CardInfo[];
  showLinkingToolbar: boolean;
  smartLinkEvents?: SmartLinkEvents;
  createAnalyticsEvent?: CreateUIAnalyticsEvent;
};

export type CardPluginOptions = CardOptions & {
  platform: 'mobile' | 'web';
  fullWidthMode?: boolean;
  createAnalyticsEvent?: CreateUIAnalyticsEvent;
  linkPicker?: LinkPickerOptions;
};

// actions
export type SetProvider = {
  type: 'SET_PROVIDER';
  provider: CardProvider | null;
};

export type Queue = {
  type: 'QUEUE';
  requests: Request[];
};

export type Resolve = {
  type: 'RESOLVE';
  url: string;
};

export type Register = {
  type: 'REGISTER';
  info: CardInfo;
};

export type ShowLinkToolbar = {
  type: 'SHOW_LINK_TOOLBAR';
};

export type HideLinkToolbar = {
  type: 'HIDE_LINK_TOOLBAR';
};

export type RegisterSmartCardEvents = {
  type: 'REGISTER_EVENTS';
  smartLinkEvents: SmartLinkEvents;
};

export type CardPluginAction =
  | SetProvider
  | Queue
  | Resolve
  | Register
  | ShowLinkToolbar
  | HideLinkToolbar
  | RegisterSmartCardEvents;

export type CardReplacementInputMethod =
  | INPUT_METHOD.CLIPBOARD
  | INPUT_METHOD.AUTO_DETECT
  | INPUT_METHOD.FORMATTING
  | INPUT_METHOD.MANUAL
  | INPUT_METHOD.TYPEAHEAD;
