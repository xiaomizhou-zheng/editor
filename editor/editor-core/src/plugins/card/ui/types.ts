import { CardAppearance } from '@atlaskit/editor-common/provider-factory';

import { Command } from '../../../types';

export interface OptionConfig {
  appearance?: CardAppearance;
  title: string;
  onClick: Command;
  selected: boolean;
  testId: string;
  disabled?: boolean;
  hidden?: boolean;
  tooltip?: string;
}
