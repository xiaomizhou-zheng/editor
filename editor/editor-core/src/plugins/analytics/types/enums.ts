export {
  EVENT_TYPE,
  ACTION,
  INPUT_METHOD,
  TRIGGER_METHOD,
  ACTION_SUBJECT,
  ACTION_SUBJECT_ID,
  FLOATING_CONTROLS_TITLE,
  CONTENT_COMPONENT,
} from '@atlaskit/editor-common/analytics';
