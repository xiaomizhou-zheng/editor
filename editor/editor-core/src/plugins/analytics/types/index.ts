export {
  ACTION,
  ACTION_SUBJECT,
  ACTION_SUBJECT_ID,
  EVENT_TYPE,
  INPUT_METHOD,
  TRIGGER_METHOD,
  CONTENT_COMPONENT,
  FLOATING_CONTROLS_TITLE,
} from './enums';
export type {
  AnalyticsDispatch,
  AnalyticsEventPayload,
  AnalyticsEventPayloadWithChannel,
  AnalyticsEventPayloadCallback,
  ErrorEventPayload,
  ErrorEventAttributes,
} from './events';
export {
  BROWSER_FREEZE_INTERACTION_TYPE,
  FULL_WIDTH_MODE,
  MODE,
  PLATFORMS,
} from './general-events';
export type { GeneralEventPayload } from './general-events';
export type {
  MediaAltTextActionType,
  MediaEventPayload,
  MediaLinkAEP,
} from './media-events';
export type { DispatchAnalyticsEvent } from './dispatch-analytics-event';
export type {
  CreateLinkInlineDialogActionType,
  CreateLinkInlineDialogEventPayload,
} from './link-tool-bar-events';

export type {
  FormatEventPayload,
  CommonListAnalyticsAttributes,
  InputMethodInsertLink,
  InputMethodInsertMedia,
  InsertEventPayload,
  LINK_REPRESENTATION,
  LINK_RESOURCE,
  LINK_STATUS,
  ListEventPayload,
  PasteContent,
  PasteEventPayload,
  PasteSource,
  PasteType,
  PasteContents,
  SubstituteEventPayload,
  TableEventPayload,
  PasteSources,
  PasteTypes,
  PASTE_ACTION_SUBJECT_ID,
} from '@atlaskit/editor-common/analytics';

export {
  USER_CONTEXT,
  DELETE_DIRECTION,
  INDENT_DIRECTION,
  PUNC,
  TABLE_ACTION,
  INDENT_TYPE,
  LIST_TEXT_SCENARIOS,
  SYMBOL,
  TABLE_BREAKOUT,
} from '@atlaskit/editor-common/analytics';
