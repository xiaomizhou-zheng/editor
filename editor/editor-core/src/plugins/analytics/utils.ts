import { Step } from 'prosemirror-transform';
import {
  EditorState,
  ReadonlyTransaction,
  Transaction,
} from 'prosemirror-state';
import { CreateUIAnalyticsEvent } from '@atlaskit/analytics-next';
import { AnalyticsStep } from '@atlaskit/adf-schema/steps';
import { editorAnalyticsChannel } from './consts';
import {
  AnalyticsEventPayloadWithChannel,
  AnalyticsEventPayload,
} from './types';
import { analyticsPluginKey } from './plugin-key';
import { HigherOrderCommand } from '../../types/command';

function getCreateUIAnalyticsEvent(
  editorState: EditorState,
): CreateUIAnalyticsEvent | null | undefined {
  return analyticsPluginKey.getState(editorState)?.createAnalyticsEvent;
}

import { attachPayloadIntoTransaction } from '../../analytics-api/attach-payload-into-transaction';
import { mapActionSubjectIdToAttributes } from '../../analytics-api/map-attributes';
import {
  getStateContext,
  getSelectionType,
  findInsertLocation,
} from '../../analytics-api/editor-state-context';

export {
  getStateContext,
  mapActionSubjectIdToAttributes,
  getSelectionType,
  findInsertLocation,
};

export function addAnalytics(
  state: EditorState,
  tr: Transaction,
  payload: AnalyticsEventPayload,
  channel: string = editorAnalyticsChannel,
): Transaction {
  const createAnalyticsEvent = getCreateUIAnalyticsEvent(state);
  if (!createAnalyticsEvent) {
    return tr;
  }

  attachPayloadIntoTransaction({
    tr,
    editorState: state,
    payload,
    channel,
  });

  return tr;
}

export type AnalyticsEventPayloadCallback = (
  state: EditorState,
) => AnalyticsEventPayload | undefined;

export function withAnalytics(
  payload: AnalyticsEventPayload | AnalyticsEventPayloadCallback,
  channel?: string,
): HigherOrderCommand {
  return (command) => (state, dispatch, view) =>
    command(
      state,
      (tr) => {
        if (dispatch) {
          if (payload instanceof Function) {
            const dynamicPayload = payload(state);
            if (dynamicPayload) {
              dispatch(addAnalytics(state, tr, dynamicPayload, channel));
            }
          } else {
            dispatch(addAnalytics(state, tr, payload, channel));
          }
        }
      },
      view,
    );
}

export function getAnalyticsEventsFromTransaction(
  tr: Transaction | ReadonlyTransaction,
): AnalyticsEventPayloadWithChannel[] {
  return (tr.steps as Step[])
    .filter<AnalyticsStep<AnalyticsEventPayload>>(
      (step: Step): step is AnalyticsStep<AnalyticsEventPayload> =>
        step instanceof AnalyticsStep,
    )
    .reduce<AnalyticsEventPayloadWithChannel[]>(
      (acc, step) => [...acc, ...step.analyticsEvents],
      [],
    );
}
