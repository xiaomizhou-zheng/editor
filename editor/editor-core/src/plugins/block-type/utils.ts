import { Node as PMNode } from 'prosemirror-model';
import { WRAPPER_BLOCK_TYPES } from './types';

export const isNodeAWrappingBlockNode = (node?: PMNode | null) => {
  if (!node) {
    return false;
  }
  return WRAPPER_BLOCK_TYPES.some(
    (blockNode) => blockNode.name === node.type.name,
  );
};
