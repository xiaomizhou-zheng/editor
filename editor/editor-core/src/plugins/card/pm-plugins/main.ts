import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import { EditorState } from 'prosemirror-state';
import { EditorView } from 'prosemirror-view';
import rafSchedule from 'raf-schd';

import { getInlineNodeViewProducer } from '../../../nodeviews/getInlineNodeViewProducer';

import { CardPluginOptions, CardPluginState } from '../types';
import reducer from './reducers';
import { PMPluginFactoryParams } from '../../../types';
import { pluginKey } from './plugin-key';
import { handleProvider, resolveWithProvider } from './util/resolve';
import {
  getNewRequests,
  getPluginState,
  getPluginStateWithUpdatedPos,
} from './util/state';
import { OutstandingRequests } from '../types';
import { EmbedCard, EmbedCardNodeViewProps } from '../nodeviews/embedCard';
import { BlockCard, BlockCardNodeViewProps } from '../nodeviews/blockCard';
import { InlineCardNodeView } from '../nodeviews/inlineCard';
import { ProviderHandler } from '@atlaskit/editor-common/provider-factory';
import { ACTION, ACTION_SUBJECT, EVENT_TYPE } from '../../../plugins/analytics';

export { pluginKey } from './plugin-key';

export const createPlugin = (options: CardPluginOptions) => (
  pmPluginFactoryParams: PMPluginFactoryParams,
) => {
  const {
    platform,
    allowResizing,
    useAlternativePreloader,
    fullWidthMode,
    createAnalyticsEvent,
  } = options;
  return new SafePlugin({
    state: {
      init(): CardPluginState {
        const {
          viewChangingExperimentToolbarStyle,
        } = pmPluginFactoryParams.featureFlags;

        pmPluginFactoryParams.dispatchAnalyticsEvent({
          eventType: EVENT_TYPE.OPERATIONAL,
          action: ACTION.EXPOSED,
          actionSubject: ACTION_SUBJECT.FEATURE,
          attributes: {
            flagKey:
              'confluence.frontend.fabric.editor.view-changing-experiment-toolbar-style',
            value: viewChangingExperimentToolbarStyle || 'noChange',
          },
          source: '@atlaskit/feature-flag-client',
          tags: ['measurement'],
        });

        return {
          requests: [],
          provider: null,
          cards: [],
          showLinkingToolbar: false,
          smartLinkEvents: undefined,
          createAnalyticsEvent,
        };
      },

      apply(tr, pluginState: CardPluginState) {
        // Update all the positions of outstanding requests and
        // cards in the plugin state.
        const pluginStateWithUpdatedPos = getPluginStateWithUpdatedPos(
          pluginState,
          tr,
        );
        // apply any actions
        const meta = tr.getMeta(pluginKey);
        if (meta) {
          const nextPluginState = reducer(pluginStateWithUpdatedPos, meta);
          return nextPluginState;
        }

        return pluginStateWithUpdatedPos;
      },
    },

    view(view: EditorView) {
      const outstandingRequests: OutstandingRequests = {};
      const subscriptionHandler: ProviderHandler<'cardProvider'> = (
        name,
        provider,
      ) => handleProvider(name, provider, view);
      const rafCancellationCallbacks: Function[] = [];

      pmPluginFactoryParams.providerFactory.subscribe(
        'cardProvider',
        subscriptionHandler,
      );

      return {
        update(view: EditorView, prevState: EditorState) {
          const currentState = getPluginState(view.state);
          const oldState = getPluginState(prevState);

          if (currentState && currentState.provider) {
            // Find requests in this state that weren't in the old one.
            const newRequests = getNewRequests(oldState, currentState);
            // Ask the CardProvider to resolve all new requests.
            const { provider } = currentState;
            newRequests.forEach((request) => {
              /**
               * Queue each asynchronous resolve request on separate frames.
               * ---
               * NB: The promise for each request is queued to take place on separate animation frames. This avoids
               * the scenario debugged and discovered in EDM-668, wherein the queuing of too many promises in quick succession
               * leads to the browser's macrotask queue being overwhelmed, locking interactivity of the browser tab.
               * By using this approach, the browser is free to schedule the resolution of the promises below in between rendering/network/
               * other tasks as per common implementations of the JavaScript event loop in browsers.
               */
              const invoke = rafSchedule(() =>
                resolveWithProvider(
                  view,
                  outstandingRequests,
                  provider,
                  request,
                  options,
                ),
              );
              rafCancellationCallbacks.push(invoke.cancel);
              invoke();
            });
          }
        },

        destroy() {
          // Cancel all outstanding requests
          Object.keys(outstandingRequests).forEach((url) =>
            Promise.reject(outstandingRequests[url]),
          );

          // Cancel any outstanding raf callbacks.
          rafCancellationCallbacks.forEach((cancellationCallback) =>
            cancellationCallback(),
          );

          pmPluginFactoryParams.providerFactory.unsubscribe(
            'cardProvider',
            subscriptionHandler,
          );
        },
      };
    },

    props: {
      nodeViews: {
        inlineCard: getInlineNodeViewProducer({
          pmPluginFactoryParams,
          Component: InlineCardNodeView,
          extraComponentProps: { useAlternativePreloader },
        }),
        blockCard: (node, view, getPos) => {
          const { portalProviderAPI, eventDispatcher } = pmPluginFactoryParams;
          const reactComponentProps: BlockCardNodeViewProps = {
            platform,
          };
          const hasIntlContext = true;
          return new BlockCard(
            node,
            view,
            getPos,
            portalProviderAPI,
            eventDispatcher,
            reactComponentProps,
            undefined,
            true,
            undefined,
            hasIntlContext,
          ).init();
        },
        embedCard: (node, view, getPos) => {
          const {
            portalProviderAPI,
            eventDispatcher,
            dispatchAnalyticsEvent,
          } = pmPluginFactoryParams;
          const reactComponentProps: EmbedCardNodeViewProps = {
            eventDispatcher,
            allowResizing,
            platform,
            fullWidthMode,
            dispatchAnalyticsEvent,
          };
          const hasIntlContext = true;
          return new EmbedCard(
            node,
            view,
            getPos,
            portalProviderAPI,
            eventDispatcher,
            reactComponentProps,
            undefined,
            true,
            undefined,
            hasIntlContext,
          ).init();
        },
      },
    },

    key: pluginKey,
  });
};
