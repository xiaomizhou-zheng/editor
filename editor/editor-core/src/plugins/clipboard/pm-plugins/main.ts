import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import { EditorState, TextSelection } from 'prosemirror-state';
import { EditorView } from 'prosemirror-view';
import { PMPluginFactoryParams } from '../../../types';
import { clipboardPluginKey } from '../plugin-key';
import { DispatchAnalyticsEvent } from '../../analytics/types/dispatch-analytics-event';
import { AnalyticsEventPayload } from '../../../plugins/analytics';
import {
  getNodeSelectionAnalyticsPayload,
  getAllSelectionAnalyticsPayload,
  getRangeSelectionAnalyticsPayload,
  getCellSelectionAnalyticsPayload,
} from '../../selection/utils';
import {
  EVENT_TYPE,
  ACTION,
  ACTION_SUBJECT,
  ACTION_SUBJECT_ID,
} from '../../analytics/types/enums';
import {
  SelectNodeAEP,
  SelectRangeAEP,
  SelectCellAEP,
} from '../../analytics/types/selection-events';
import { DOMSerializer } from 'prosemirror-model';
import { findParentNodeOfType } from 'prosemirror-utils';
import { Fragment } from 'prosemirror-model';
import type { NodeType, Schema } from 'prosemirror-model';

export const createPlugin = ({
  dispatchAnalyticsEvent,
  schema,
}: PMPluginFactoryParams) => {
  let editorView: EditorView;
  const getEditorView = () => editorView;
  return new SafePlugin({
    key: clipboardPluginKey,
    view: (view: EditorView) => {
      editorView = view;
      return {
        update: (view: EditorView) => {
          editorView = view;
        },
      };
    },
    props: {
      handleDOMEvents: {
        cut: (view) =>
          sendClipboardAnalytics(view, dispatchAnalyticsEvent, ACTION.CUT),
        copy: (view) =>
          sendClipboardAnalytics(view, dispatchAnalyticsEvent, ACTION.COPIED),
      },
      clipboardSerializer: createClipboardSerializer(schema, getEditorView),
    },
  });
};

/**
 * Overrides Prosemirror's default clipboardSerializer, in order to fix table row copy/paste bug raised in ED-13003.
 * This allows us to store the original tableâ€™s attributes on the new table that the row is wrapped with when it is being copied.
 * e.g. keeping the layout on a row that is copied.
 * We store the default serializer in order to call it after we handle the table row case.
 */
export const createClipboardSerializer = (
  schema: Schema,
  getEditorView: () => EditorView,
): DOMSerializer => {
  const oldSerializer = DOMSerializer.fromSchema(schema);
  const newSerializer = new DOMSerializer(
    oldSerializer.nodes,
    oldSerializer.marks,
  );

  const originalSerializeFragment = newSerializer.serializeFragment.bind(
    newSerializer,
  );

  newSerializer.serializeFragment = (
    content: Fragment,
    options: { [key: string]: any } = {},
    target?: Node,
  ): DocumentFragment => {
    const editorView = getEditorView();
    const selection = editorView.state.selection;

    // We do not need to handle when a user copies a tableRow + other content.
    // In that scenario it already wraps the Row with correct Table and attributes.
    if (!options.tableWrapperExists) {
      let i = 0;
      while (i < content.childCount) {
        if (content.maybeChild(i)?.type.name === 'table') {
          options.tableWrapperExists = true;
          break;
        }
        i++;
      }
    }

    // When the content being copied includes a tableRow that is not already wrapped with a table,
    // We will wrap it with one ourselves, while preserving the parent table's attributes.
    if (
      content.firstChild?.type.name === 'tableRow' &&
      !options.tableWrapperExists
    ) {
      // We only want 1 table wrapping the rows.
      // tableWrapperExist is a custom prop added solely for the purposes of this recursive algorithm.
      // The function is recursively called for each node in the tree captured in the fragment.
      // For recursive logic see the bind call above and the prosemirror-model (https://github.com/ProseMirror/prosemirror-model/blob/master/src/to_dom.js#L44
      // and https://github.com/ProseMirror/prosemirror-model/blob/master/src/to_dom.js#L87)
      options.tableWrapperExists = true;

      const parentTable = findParentNodeOfType(schema.nodes.table)(selection);
      const attributes = parentTable?.node.attrs;

      const newTable: NodeType = schema.nodes.table;
      const newTableNode = newTable.createChecked({ ...attributes }, content);
      const newContent = Fragment.from(newTableNode);
      // Pass updated content into original ProseMirror serializeFragment function.
      // Currently incorrectly typed in @Types. See this GitHub thread: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57668
      //@ts-ignore
      return originalSerializeFragment(newContent, options, target);
    }

    // If we're not copying any rows, just run default serializeFragment function.
    // Currently incorrectly typed in @Types. See this GitHub thread: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57668
    //@ts-ignore
    return originalSerializeFragment(content, options, target);
  };
  return newSerializer;
};

export const sendClipboardAnalytics = (
  view: EditorView,
  dispatchAnalyticsEvent: DispatchAnalyticsEvent,
  action: ACTION.CUT | ACTION.COPIED,
) => {
  const clipboardAnalyticsPayload = getAnalyticsPayload(view.state, action);
  if (clipboardAnalyticsPayload) {
    dispatchAnalyticsEvent(clipboardAnalyticsPayload);
  }
  // return false so we don't block any other plugins' cut or copy handlers
  // from running just because we are sending an analytics event
  return false;
};

export const getAnalyticsPayload = (
  state: EditorState,
  action: ACTION.CUT | ACTION.COPIED,
): AnalyticsEventPayload | undefined => {
  const { selection, doc } = state;
  const selectionAnalyticsPayload =
    getNodeSelectionAnalyticsPayload(selection) ||
    getRangeSelectionAnalyticsPayload(selection, doc) ||
    getAllSelectionAnalyticsPayload(selection) ||
    getCellSelectionAnalyticsPayload(state);

  if (selectionAnalyticsPayload) {
    const {
      actionSubjectId: selectionActionSubjectId,
    } = selectionAnalyticsPayload;

    let content: string[] = [];
    switch (selectionActionSubjectId) {
      case ACTION_SUBJECT_ID.NODE:
        content.push(
          (selectionAnalyticsPayload as SelectNodeAEP).attributes!.node,
        );
        break;
      case ACTION_SUBJECT_ID.RANGE:
        content.push(
          ...(selectionAnalyticsPayload as SelectRangeAEP).attributes!.nodes,
        );
        break;
      case ACTION_SUBJECT_ID.ALL:
        content.push('all');
        break;
      case ACTION_SUBJECT_ID.CELL: {
        const {
          selectedCells,
        } = (selectionAnalyticsPayload as SelectCellAEP).attributes!;
        content.push(...Array(selectedCells).fill('tableCell'));
        break;
      }
    }

    return {
      eventType: EVENT_TYPE.TRACK,
      action,
      actionSubject: ACTION_SUBJECT.DOCUMENT,
      attributes: {
        content,
      },
    };
  }

  if (selection instanceof TextSelection && selection.$cursor) {
    return {
      eventType: EVENT_TYPE.TRACK,
      action,
      actionSubject: ACTION_SUBJECT.DOCUMENT,
      attributes: {
        content: ['caret'],
      },
    };
  }
};
