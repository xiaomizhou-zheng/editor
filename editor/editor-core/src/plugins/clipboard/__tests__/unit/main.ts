import { PluginKey } from 'prosemirror-state';
import { EditorView } from 'prosemirror-view';
import { clipboardPluginKey } from './../../plugin-key';
import {
  createProsemirrorEditorFactory,
  LightEditorPlugin,
  Preset,
  DispatchAnalyticsEvent,
} from '@atlaskit/editor-test-helpers/create-prosemirror-editor';
import {
  doc,
  hr,
  p,
  table,
  tr,
  th,
  td,
  DocBuilder,
} from '@atlaskit/editor-test-helpers/doc-builder';
import rulePlugin from '../../../rule';
import layoutPlugin from '../../../layout';
import { tablesPlugin } from '@atlaskit/editor-plugin-table';
import clipboardPlugin from '../../index';
import { sendClipboardAnalytics } from '../../pm-plugins/main';
import { ACTION } from '../../../analytics/types/enums';
import {
  setNodeSelection,
  setTextSelection,
  setAllSelection,
  setCellSelection,
} from '../../../../utils/selection';
// @ts-ignore
import { __serializeForClipboard } from 'prosemirror-view';
import { selectRow } from '@atlaskit/editor-tables/src/utils';
import { TableAttributes } from '@atlaskit/adf-schema';

const TABLE_LOCAL_ID = 'test-table-local-id';

describe('clipboard plugin', () => {
  const createEditor = createProsemirrorEditorFactory();
  const preset = new Preset<LightEditorPlugin>()
    .add(clipboardPlugin)
    .add(rulePlugin)
    .add(layoutPlugin)
    .add([
      tablesPlugin,
      {
        tableOptions: {
          allowNumberColumn: true,
          allowHeaderRow: true,
          allowHeaderColumn: true,
          allowBackgroundColor: true,
          permittedLayouts: 'all',
        },
      },
    ]);

  const editor = (doc: DocBuilder) =>
    createEditor<undefined, PluginKey>({
      doc,
      preset,
      pluginKey: clipboardPluginKey,
    });

  let editorView: EditorView;
  let refs: { [name: string]: number };
  let dispatchAnalyticsEvent: DispatchAnalyticsEvent;

  describe('analytics', () => {
    const analyticsTests = (actionType: 'cut' | 'copied') => {
      const actionTypeEnum = actionType === 'cut' ? ACTION.CUT : ACTION.COPIED;

      beforeEach(() => {
        ({ editorView, refs, dispatchAnalyticsEvent } = editor(
          doc(
            '{pStart}',
            p('testing clipboard plugin{<>}'),
            '{ruleStart}',
            hr(),
            '{ruleEnd}',
          ),
        ));
        (dispatchAnalyticsEvent as jest.Mock).mockClear();
      });

      it(`fires analytics event when node selection is ${actionType}`, () => {
        setNodeSelection(editorView, refs.ruleStart);
        sendClipboardAnalytics(
          editorView,
          dispatchAnalyticsEvent,
          actionTypeEnum,
        );
        expect(dispatchAnalyticsEvent).toHaveBeenCalledWith({
          eventType: 'track',
          action: actionType,
          actionSubject: 'document',
          attributes: {
            content: ['rule'],
          },
        });
      });

      it(`fires analytics event when all selection is ${actionType}`, () => {
        setAllSelection(editorView);
        sendClipboardAnalytics(
          editorView,
          dispatchAnalyticsEvent,
          actionTypeEnum,
        );
        expect(dispatchAnalyticsEvent).toHaveBeenCalledWith({
          eventType: 'track',
          action: actionType,
          actionSubject: 'document',
          attributes: {
            content: ['all'],
          },
        });
      });

      it(`fires analytics event when range selection is ${actionType}`, () => {
        setTextSelection(editorView, refs.pStart, refs.ruleEnd);
        sendClipboardAnalytics(
          editorView,
          dispatchAnalyticsEvent,
          actionTypeEnum,
        );
        expect(dispatchAnalyticsEvent).toHaveBeenCalledWith({
          eventType: 'track',
          action: actionType,
          actionSubject: 'document',
          attributes: {
            content: ['paragraph', 'rule'],
          },
        });
      });

      it(`fires analytics event when empty selection is ${actionType}`, () => {
        setTextSelection(editorView, refs.pStart);
        sendClipboardAnalytics(
          editorView,
          dispatchAnalyticsEvent,
          actionTypeEnum,
        );
        expect(dispatchAnalyticsEvent).toHaveBeenCalledWith({
          eventType: 'track',
          action: actionType,
          actionSubject: 'document',
          attributes: {
            content: ['caret'],
          },
        });
      });

      it(`fires analytics event when cell selection is ${actionType}`, () => {
        ({ editorView, refs, dispatchAnalyticsEvent } = editor(
          doc(
            table()(
              tr(th()(p('{<>}a1')), '{cell2}', th()(p('a2'))),
              tr(td()(p('b1')), '{cell4}', td()(p('b2'))),
            ),
          ),
        ));
        setCellSelection(editorView, refs.cell2, refs.cell4);
        sendClipboardAnalytics(
          editorView,
          dispatchAnalyticsEvent,
          actionTypeEnum,
        );
        expect(dispatchAnalyticsEvent).toHaveBeenCalledWith({
          eventType: 'track',
          action: actionType,
          actionSubject: 'document',
          attributes: {
            content: ['tableCell', 'tableCell'],
          },
        });
      });
    };

    describe('cut', () => {
      analyticsTests('cut');
    });

    describe('copy', () => {
      analyticsTests('copied');
    });
  });

  describe('clipboardSerializer', () => {
    describe('when copying a table row from a table', () => {
      it.each(['default', 'wide', 'full-width'])(
        'should write to the clipboard a table keeping the same attributes, including the layout',
        (tableLayout) => {
          const { editorView } = editor(
            doc(
              table({
                layout: tableLayout as TableAttributes['layout'],
                localId: TABLE_LOCAL_ID,
              })(
                tr(th()(p('1')), th()(p('2')), th()(p('3'))),
                tr(td()(p('4')), td()(p('5')), td()(p('6'))),
                tr(td()(p('7')), td()(p('8')), td()(p('9'))),
              ),
            ),
          );

          // select the row
          const rowSelector = selectRow(0);
          const newTr = rowSelector(editorView.state.tr);

          editorView.dispatch(newTr);

          const { dom } = __serializeForClipboard(
            editorView,
            editorView.state.selection.content(),
          );

          expect(
            dom.querySelector('table').getAttribute('data-layout'),
          ).toEqual(tableLayout);

          // Snapshotting to ensure table and all attributes wrap the row and are written to clipboard.
          expect(dom).toMatchSnapshot();
        },
      );
    });
  });
});
