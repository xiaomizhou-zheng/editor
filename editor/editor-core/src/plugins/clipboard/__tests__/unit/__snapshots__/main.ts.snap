// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clipboard plugin clipboardSerializer when copying a table row from a table should write to the clipboard a table keeping the same attributes, including the layout 1`] = `
<div>
  <table
    data-autosize="false"
    data-layout="default"
    data-number-column="false"
    data-pm-slice="1 1 []"
    data-table-local-id="test-table-local-id"
  >
    <tbody>
      <tr>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            1
          </p>
        </th>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            2
          </p>
        </th>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            3
          </p>
        </th>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`clipboard plugin clipboardSerializer when copying a table row from a table should write to the clipboard a table keeping the same attributes, including the layout 2`] = `
<div>
  <table
    data-autosize="false"
    data-layout="wide"
    data-number-column="false"
    data-pm-slice="1 1 []"
    data-table-local-id="test-table-local-id"
  >
    <tbody>
      <tr>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            1
          </p>
        </th>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            2
          </p>
        </th>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            3
          </p>
        </th>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`clipboard plugin clipboardSerializer when copying a table row from a table should write to the clipboard a table keeping the same attributes, including the layout 3`] = `
<div>
  <table
    data-autosize="false"
    data-layout="full-width"
    data-number-column="false"
    data-pm-slice="1 1 []"
    data-table-local-id="test-table-local-id"
  >
    <tbody>
      <tr>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            1
          </p>
        </th>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            2
          </p>
        </th>
        <th
          class="pm-table-header-content-wrap"
        >
          <p>
            3
          </p>
        </th>
      </tr>
    </tbody>
  </table>
</div>
`;
